contrib/connectors/pandas/models.py

```diff
@@ -0,0 +1,940 @@
+        d['time_grain_sqla'] = [(g, g) for g in self.GRAINS.keys()]
+        logging.info(d)
+        return d
+
+    @property
+    def pandas_read_method(self):
+        try:
+            # The format is a Choice object
+            format = self.format.code
+        except AttributeError:
+            format = self.format
+        return getattr(pd, 'read_{format}'.format(format=format))
+
+    @property
+    def pandas_read_parameters(self):
+        return self.additional_parameters or {}
+
+    def get_empty_dataframe(self):
+        """Create an empty dataframe with the correct columns and dtypes"""
+        columns = []
+        for col in self.columns:
+            type = ('datetime64[ns]'
+                    if is_datetime64_any_dtype(col.type)
+                    else col.type)
+            columns.append((col.column_name, type))
+        return pd.DataFrame({k: pd.Series(dtype=t) for k, t in columns})
+
+    @property
+    def cache_key(self):
+        source = {'source_url': self.source_url,
+                  'source_auth': self.source_auth}
+        source.update(self.source_parameters or {})
+        source.update(self.pandas_read_parameters)
+        s = str([(k, source[k]) for k in sorted(source.keys())])
+        return hashlib.md5(s.encode('utf-8')).hexdigest()
+
+    def get_dataframe(self):
+        """
+        Read the source_url and return a Pandas DataFrame.
+
+        Use the PandasColumns to coerce columns into the correct dtype,
+        and add any calculated columns to the DataFrame.
+        """
+        if self.df is None:
+            if dataframe_cache:
+                cache_key = self.cache_key
+                self.df = dataframe_cache.get(cache_key)
+            if not isinstance(self.df, pd.DataFrame):
+                if (isinstance(self.source_url, basestring) and
+                        self.source_url[:4] == 'http'):
+                    # Use requests to retrieve remote data so we can handle authentication
+                    auth = self.source_auth
+                    url = self.source_url
+                    if isinstance(auth, (tuple, list)):
+                        response = requests.get(url, params=self.source_parameters,
+                                                auth=tuple(auth))
+                    elif auth:
+                        response = requests.get(url, params=self.source_parameters,
+                                                headers={'Authorization': auth})
+                    else:
+                        response = requests.get(url, params=self.source_parameters)
+                    response.raise_for_status()
+                    data = BytesIO(response.content)
+                else:
+                    # Local file, so just use Pandas directly
+                    data = self.source_url
+                # Read the dataframe from the response
+                self.df = self.pandas_read_method(data, **self.pandas_read_parameters)
```

Comment:
I think adding a blank line before this make it  more readable as this is not part of else. However I am not sure about lint rules here

However I do see blank line used to make code readable and communicate thsi line is in parent block

Reply:
Not taking care these things in current PR, may be a separate US we can check the same.

Reply:
ok. I  was more interested the knowing.  We can follow where it make sense